ssl/quic/qlog_event_helpers.c

```diff
@@ -0,0 +1,633 @@
+                QLOG_END();
+            QLOG_END();
+        }
+        break;
+    default:
+unknown:
+        QLOG_STR("frame_type", "unknown");
+        QLOG_U64("frame_type_value", frame_type);
+
+        /*
+         * Can't continue scanning for frames in this case as the frame length
+         * is unknown. We log the entire body the rest of the packet payload
+         * as the raw data of the frame.
+         */
+        QLOG_BEGIN("raw");
+            QLOG_BIN("data", PACKET_data(&orig_pkt),
+                     PACKET_remaining(&orig_pkt));
+        QLOG_END();
+        ignore_res(PACKET_forward(pkt, PACKET_remaining(pkt)));
+        break;
+    }
+
+    return 1;
+}
+
+static void log_frame(QLOG *qlog_instance, PACKET *pkt,
+                      uint64_t *need_skip)
+{
+    size_t rem_before, rem_after;
+
+    rem_before = PACKET_remaining(pkt);
+
+    if (!log_frame_actual(qlog_instance, pkt, need_skip))
+        return;
+
+    rem_after = PACKET_remaining(pkt);
+    QLOG_U64("length", rem_before - rem_after);
+}
+
+static int log_frames(QLOG *qlog_instance,
+                      const OSSL_QTX_IOVEC *iovec,
+                      size_t num_iovec)
+{
+    size_t i;
+    PACKET pkt;
+    uint64_t need_skip = 0;
+
+    for (i = 0; i < num_iovec; ++i) {
+        if (!PACKET_buf_init(&pkt, iovec[i].buf, iovec[i].buf_len))
+            return 0;
+
+        while (PACKET_remaining(&pkt) > 0) {
+            if (need_skip > 0) {
+                size_t adv = need_skip;
+
+                if (adv < PACKET_remaining(&pkt))
+                    adv = PACKET_remaining(&pkt);
+
+                if (!PACKET_forward(&pkt, adv))
+                    return 0;
+
+                need_skip -= adv;
+                continue;
+            }
+
+            QLOG_BEGIN(NULL)
+            {
+                log_frame(qlog_instance, &pkt, &need_skip);
+            }
+            QLOG_END()
+        }
+    }
+
+    return 1;
+}
+
+static void log_packet(QLOG *qlog_instance,
+                       const QUIC_PKT_HDR *hdr,
+                       QUIC_PN pn,
+                       const OSSL_QTX_IOVEC *iovec,
+                       size_t num_iovec,
+                       uint64_t datagram_id)
+{
+    const char *type_s;
+
+    QLOG_BEGIN("header")
+        type_s = quic_pkt_type_to_qlog(hdr->type);
+        if (type_s == NULL)
+            type_s = "unknown";
+
+        QLOG_STR("packet_type", type_s);
+        if (ossl_quic_pkt_type_has_pn(hdr->type))
+            QLOG_U64("packet_number", pn);
+
+        QLOG_CID("dcid", &hdr->dst_conn_id);
+        if (ossl_quic_pkt_type_has_scid(hdr->type))
+            QLOG_CID("scid", &hdr->src_conn_id);
+
+        if (hdr->token_len > 0) {
+            QLOG_BEGIN("token")
+                QLOG_BEGIN("raw")
+                    QLOG_BIN("data", hdr->token, hdr->token_len);
+                QLOG_END()
+            QLOG_END()
+        }
+        /* TODO(QLOG): flags, length */
```

Comment:
What is the plan to resolve thsi?

Reply:
Inessential feature which can be added in a future PR. I'll mark these QLOG FUTURE if that's convenient.

Reply:
Please do. Or change it to `/* Possible enhancement: handle flags, length */